backend/de.metas.monitoring/src/main/java/de/metas/monitoring/adapter/MicrometerPerformanceMonitoringService.java

```diff
@@ -42,63 +42,88 @@
-			@NonNull final Optional<APMPerformanceMonitoringService> apmPerformanceMonitoringService,
 			@NonNull final MeterRegistry meterRegistry)
 	{
-		this.apmPerformanceMonitoringService = apmPerformanceMonitoringService;
 		this.meterRegistry = meterRegistry;
 	}
 
 	@Override
-	public <V> V monitorTransaction(
+	public <V> V monitor(
 			@NonNull final Callable<V> callable,
-			@NonNull final TransactionMetadata metadata)
+			@NonNull final PerformanceMonitoringService.Metadata metadata)
 	{
-		final ArrayList<Tag> tags = createTagsFromLabels(metadata.getLabels());
-		mkTagIfNotNull("Name", metadata.getName()).ifPresent(tags::add);
 
-		final Callable<V> callableToUse;
-		if (apmPerformanceMonitoringService.isPresent())
+		final PerformanceMonitoringData perfMonData = perfMonDataTL.get();
+		final ArrayList<Tag> tags = new ArrayList<>();
+
+		mkTagIfNotNull("name", metadata.getClassName()).ifPresent(tags::add);
+		mkTagIfNotNull("action", metadata.getFunctionName()).ifPresent(tags::add);
+
+		if(perfMonData.getDepth() == 0
+				&& metadata.getType() == Type.REST_CONTROLLER_WITH_WINDOW_ID
+				&& metadata.getFunctionName() != null
+				&& metadata.getWindowNameAndId() != null)
 		{
-			callableToUse = () -> apmPerformanceMonitoringService.get().monitorTransaction(callable, metadata);
+			perfMonData.setIsInitiatorLabelActive(true);
+			perfMonData.setInitiator(metadata.getClassName() + " - " + metadata.getFunctionName());
+			perfMonData.setInitiatorWindow(metadata.getWindowNameAndId());
+			perfMonData.getCalledBy().add(0, "HTTP Request");
+
 		}
-		else
+		else if(perfMonData.getDepth() == 0
+				&& metadata.getType() == Type.REST_CONTROLLER
+				&& metadata.getFunctionName() != null)
 		{
-			callableToUse = callable;
+			perfMonData.setIsInitiatorLabelActive(true);
+			perfMonData.setInitiator(metadata.getClassName() + " - " + metadata.getFunctionName());
+			perfMonData.setInitiatorWindow("NONE");
+			perfMonData.getCalledBy().add(0, "HTTP Request");
 		}
-		return recordCallable(callableToUse, tags, "mf." + metadata.getType().getCode());
-	}
-
-	@Override
-	public <V> V monitorSpan(
-			@NonNull final Callable<V> callable,
-			@NonNull final SpanMetadata metadata)
-	{
-		final ArrayList<Tag> tags = createTagsFromLabels(metadata.getLabels());
-
-		mkTagIfNotNull("Name", metadata.getName()).ifPresent(tags::add);
-		mkTagIfNotNull("SubType", metadata.getSubType()).ifPresent(tags::add);
-		mkTagIfNotNull("Action", metadata.getAction()).ifPresent(tags::add);
-
-		final Callable<V> callableToUse;
-		if (apmPerformanceMonitoringService.isPresent())
+		else if(perfMonData.getDepth() == 0)
 		{
-			callableToUse = () -> apmPerformanceMonitoringService.get().monitorSpan(callable, metadata);
+			perfMonData.setIsInitiatorLabelActive(false);
 		}
-		else
+
+		if(perfMonData.getIsInitiatorLabelActive())
 		{
-			callableToUse = callable;
+			mkTagIfNotNull("depth", String.valueOf(perfMonData.getDepth())).ifPresent(tags::add);
```

Comment:
i kind of don't like this construction at this level of application:

> mkTagIfNotNull("depth", String.valueOf(perfMonData.getDepth())).ifPresent(tags::add);

why not having a method called

addTagIfNotNull(ArrayList<Tag> tags, String name, String value)

which avoids creating unnecessary objects (like Optionals) 
and which simply does what it says.

NOTE: on other levels of application, using Optionals when needed is perfectly OK, but here, we must make sure our measuring overhead is MINIMAL.